package/MDAnalysis/coordinates/H5MD.py

```diff
@@ -727,3 +757,112 @@ def has_forces(self):
     @has_forces.setter
     def has_forces(self, value: bool):
         self._has['force'] = value
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        del state['_particle_group']
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__ = state
+        self._particle_group = self._file['particles'][
+                               list(self._file['particles'])[0]]
+        self[self.ts.frame]
+
+
+class H5PYPicklable(h5py.File):
+    """H5PY file object (read-only) that can be pickled.
+
+    This class provides a file-like object (as returned by
+    :class:`h5py.File`) that,
+    unlike standard Python file objects,
+    can be pickled. Only read mode is supported.
+
+    When the file is pickled, filename, mode, driver, and comm of
+    :class:`h5py.File` in the file are saved. On unpickling, the file
+    is opened by filename, mode, driver, and comm (driver and comm are
+    optional arguments). This means that for a successful unpickle, the
+    original file still has to be accessible with its filename.
+
+    Parameters
+    ----------
+    filename : str or file-like
+        a filename given a text or byte string.
+    driver : str (optional)
+        H5PY file driver used to open H5MD file
+    comm : :class:`MPI.Comm` (optional)
+        MPI communicator used to open H5MD file
+        Must be passed with `'mpio'` file driver
+
+    Example
+    -------
+    ::
+
+        f = H5PYPicklable('filename', 'r')
+        print(f['particles/trajectory/position/value'][0])
+        f.close()
+
+    can also be used as context manager::
+
+        with H5PYPicklable('filename', 'r'):
+            print(f['particles/trajectory/position/value'][0])
+
+    See Also
+    ---------
+    :class:`MDAnalysis.lib.picklable_file_io.FileIOPicklable`
+    :class:`MDAnalysis.lib.picklable_file_io.BufferIOPicklable`
+    :class:`MDAnalysis.lib.picklable_file_io.TextIOPicklable`
+    :class:`MDAnalysis.lib.picklable_file_io.GzipPicklable`
+    :class:`MDAnalysis.lib.picklable_file_io.BZ2Picklable`
+
+
+    .. versionadded:: 2.0.0
+    """
+
+    def __getstate__(self):
+        driver = self.driver
+        if driver == 'mpio':
```

Comment:
In order to move the PR along, make `__getstate__` fail (with `TypeError`) when you have driver mpio and a comm object.

Add a test that specifically checks for the exception being raised. Then open a new issue to make pickling of parallel h5py files work.

Reply:
The test fails with `ModuleNotFoundError` because h5py tries to import mpi4py when `driver='mpio'`. Is there any way around this?

Reply:
Which test are we talking about?

Reply:
something like this:
```
@pytest.mark.skipif(not HAS_H5PY, reason="h5py not installed")
def test_H5MD_parallel_pickle():
    from unittest.mock import MagicMock
    MPI_COMM_WORLD = MagicMock()
    h5md_io = H5PYPicklable(H5MD_xvf, 'r',
                            driver="mpio",
                            comm=MPI_COMM_WORLD)
    h5md_io_pickled = pickle.loads(pickle.dumps(h5md_io))
```

Reply:
Try mocking the mpi4py module, too.  The idea is to provide a thing (the mock) that just pretends to be there but doesn't really do anything.

Reply:
@richardjgowers you have been dubbed in other PR "The King of Mocks" – any suggestions?

Reply:
I assume it cannot be fixed as it's checked inside h5py code. (https://github.com/h5py/h5py/blob/8cc83410995996330c85867d0e56cbe922b709f8/h5py/_hl/files.py#L43)

Maybe just excluding these lines from coverage for now?

Reply:
There might be a way to trick `h5py` into believing that it was built with mpi support (one could patch the call to `h5.get_config()` https://github.com/h5py/h5py/blob/8cc83410995996330c85867d0e56cbe922b709f8/h5py/_hl/files.py#L24 or just the `mpi` variable) but that's not worth it for right now so I concur with @yuxuanzhuang to just skip the test. Add a comment summarizing the current issues and move forward.

Reply:
@orbeckst  just pushed commit with the lines excluded from coverage